ci/3-setup-k8s

```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+# NOTE: The script should be executed as a part of ci/run!
+
+# SCRIPT SUMMARY
+# --------------
+# 1. Start and await minikube, if required
+# 2. Initialize Helm, if required
+
+# FIXME: currently we only check if minikube+helm is ready, not first if
+#        minikube is ready and then if also helm is ready.
+
+# Set shell options
+# -e : Exit immediately if a command exits with a non-zero status.
+# -u : Treat unset variables as an error when substituting.
+# -x : Print commands and their arguments as they are executed.
+set -eux
+
+# Setup Kubernetes
+# ----------------
+
+# Don't use the default profile: this helps to make it clear whats what while
+# inspecting virtualbox's VMs.
+minikube profile z2jh
+
+if set +u && [[ ! -z "${CICD}" ]] && set -u; then
+  # CICD
+  echo "starting minikube - CICD mode"
+  minikube config set vm-driver none
+  minikube config set WantNoneDriverWarning false
+  export CHANGE_MINIKUBE_NONE_USER=true
+
+  minikube addons disable dashboard
+  SUDO='sudo -E'
+else
+  # LOCAL
+  if (timeout 3 bash -c 'kubectl get pod -n kube-system --selector app=helm,name=tiller | grep 1/1'); then
+    echo "minikube cluster was already setup"
+    echo "--- you can re-setup the cluster, remove it first with `ci/clean`"
+    exit 0
+  fi
+  echo "starting minikube - LOCAL mode"
+  minikube config set vm-driver virtualbox
+
+  SUDO=''
+fi;
+minikube config set kubernetes-version v${KUBE_VERSION}
+minikube config set WantReportErrorPrompt false
+minikube config set WantKubectlDownloadMsg false
+minikube config set ShowBootstrapperDeprecationNotification false
+
+if [[ ${KUBE_VERSION} == 1.8* ]] || [[ ${KUBE_VERSION} == 1.9* ]] || [[ ${KUBE_VERSION} == 1.10.0* ]]; then
+  # WARNING: localkube not supported beyond Kubernetes 1.10.0
+    $(echo $SUDO) $(which minikube) start --bootstrapper=localkube --extra-config=apiserver.Authorization.Mode=RBAC
+else
+  # WARNING: kubeadm requires systemd. systemd isn't available for Ubuntu 14.04
+  # (trusty) VMs as used by Travis and CircleCI. It is available on Ubuntu 16.04
+  # (xenial) or 18.04 (bionic) VMs though.
+    $(echo $SUDO) $(which minikube) start
+fi
+
+# IF LOCAL ENVIRONMENT
+if set +u && [[ -z "${CICD}" ]] && set -u; then
+  echo "update minikube context"
+  # minikube update-context
```

Comment:
Why not run this? If we want to tell people to run this locally (I think we should!) then we need to make it very hard to accidentally run the following `kubectl` commands on anything but a minikube cluster (or some other cluster the user specifically configured. The number of times I've had my kube context set incorrectly and nearly deployed/wiped something from the wrong cluster is too big for me to publicly say.

So I think we should have some simple checks that protect the tired, frustrated and/or stressed user who is running this to debug stuff from accidentally running these on anything but a minikube.

Reply:
Yes I agree!
Let's add some fail safes, it will be easy because I've already tried to make this a sandbox.

Sandbox steps:
- [x] Use $KUBECONFIG, $HELM_HOME, and $MINIKUBE_HOME with in-repo .kube/ .minikube/ .helm/
- [ ] Run checks that `which kubectl` `which minikube` and `which helm` points to the binary in repo.

Reply:
I ran into issues with the minikube update-context command, I dont think it played well with a custom .kube/config location or that I configured the minikube's profile, which will influence things i don't fully know but I think one of these things are the name of the minikube VM created for example, to avoid clashes with other VMs

minikube update-context is described to...

```
	Use:   "update-context",
	Short: "Verify the IP address of the running cluster in kubeconfig.",
	Long: `Retrieves the IP address of the running cluster, checks it
			with IP in kubeconfig, and corrects kubeconfig if incorrect.`,
```